File: algorithm/322.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <ctype.h>
#include <math.h>
#define MIN(x, y) (x < y ? x : y)
int coinChange(int *coins, int coinsSize, int amount)
{
    int *dp = (int *)calloc(amount + 1, sizeof(int));
    for (int i = 1; i < amount + 1; i++)
    {
        dp[i] = INT_MAX;
    }
    for (int i = 1; i <= amount; i++)
    {
        for (int j = 0; j < coinsSize; j++)
        {
            if (i >= coins[j])
            {
                dp[i] = MIN(dp[i], dp[i - coins[j]] + 1);
            }
        }
    }
    if (dp[amount] == INT_MAX)
    {
        return -1;
    }
    return dp[amount];
}
int main(int argc, char *argv[])
{
    int a[3] = {1, 2, 5};
    printf("%d\n", coinChange(a, 3, 11));
    return 0;
}

File: go/goalgor/algor/322.go
func Min(x, y int) int {
	if x < y {
		return x
	}
	return y
}
func coinChange(coins []int, amount int) int {
	n := len(coins)
	MaxInt := 1000000
	record := make([][]int, n+1)
	for i := 0; i < n+1; i++ {
		record[i] = make([]int, amount+1)
	}

	for j := 0; j <= n; j++ {
		for i := 0; i < amount+1; i++ {
			record[j][i] = MaxInt
		}
	}
	record[0][0] = 0

	for i := 1; i < n+1; i++ {
		val := coins[i-1]
		for j := 0; j < amount+1; j++ {
			for k := 0; k*val <= j; k++ {
				record[i][j] = Min(record[i][j], record[i-1][j-k*val]+k)
			}
		}
	}
	if record[n][amount] == MaxInt {
		return -1
	}
	return record[n][amount]
}